Problem (fix #459): RoundUp in subscription module is not benchmarked

yihuang opened the pull request:
Solution:
- benchmark several cases
- also added several unit test cases and format some code

result on my mac:

```
goos: darwin
goarch: amd64
pkg: github.com/crypto-org-chain/chain-main/v1/x/subscription/types
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkRoundUpYearly
BenchmarkRoundUpYearly-16      	 6438783	       178.3 ns/op
BenchmarkRoundUpMonthly
BenchmarkRoundUpMonthly-16     	 6354093	       195.8 ns/op
BenchmarkRoundUpWeekly
BenchmarkRoundUpWeekly-16      	 4748344	       249.1 ns/op
BenchmarkRoundUpLeapYear
BenchmarkRoundUpLeapYear-16    	16125798	        74.17 ns/op
BenchmarkRoundUpOverlap
BenchmarkRoundUpOverlap-16     	 4636618	       259.1 ns/op
BenchmarkCompile
BenchmarkCompile-16            	 2755452	       429.5 ns/op
BenchmarkCountPeriods
BenchmarkCountPeriods-16       	    1090	   1052000 ns/op
```

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻


# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

JayT106 commented on x/subscription/types/cronspec_bench_test.go:
```diff
@@ -0,0 +1,105 @@
+	compiled := spec.Compile()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		compiled.CountPeriods(0, 31536000, 0) // period of a year
```
looks like the CountPeriods can be improved in the future by counting with the binary search like algo.

yihuang replied:
> looks like the CountPeriods can be improved in the future by counting with the binary search like algo.
> 
> BTW, LGTM.

I don’t see how to do that, can you elaborate on that?
I can see there’s some significant optimization can be done to some special cases, but don’t see a generic one.

JayT106 replied:
I didn't think through the cronSpec.RoundUp function. Yes, you are right!

leejw51crypto reviewed (approved):
lgtm

tomtau reviewed (changes requested):
tests fail after the latest master merge: https://github.com/crypto-org-chain/chain-main/pull/461/checks?check_run_id=2302150025#step:13:15

yihuang commented:
> tests fail after the latest master merge: https://github.com/crypto-org-chain/chain-main/pull/461/checks?check_run_id=2302150025#step:13:15

done, rebased and fixed the module import path.

tomtau commented:
subscription integration test fails: https://github.com/crypto-org-chain/chain-main/pull/461/checks?check_run_id=2303017902#step:5:1955
